core/src/main/scala/frameless/CatalystRowOrdered.scala

```diff
@@ -0,0 +1,45 @@
+package frameless
+
+import shapeless._
+
+import scala.annotation.implicitNotFound
+
+/** Types that can be used to sort a dataset by Catalyst. */
+@implicitNotFound("Cannot order by columns of type ${A}.")
+trait CatalystRowOrdered[A]
+
+object CatalystRowOrdered {
+  private[this] val theInstance = new CatalystRowOrdered[Any] {}
+  private[this] def of[A]: CatalystRowOrdered[A] = theInstance.asInstanceOf[CatalystRowOrdered[A]]
+
+  /*
+  The following are sortable by spark:
+  see [[org.apache.spark.sql.catalyst.expressions.RowOrdering.isOrderable]]
+    AtomicType
+    StructType containing only orderable types
+    ArrayType containing only orderable types
+    UserDefinedType containing only orderable types
+
+  MapType can't be used in order!
+  TODO: UDF
+   */
+
+  implicit def orderedEvidence[A](implicit catalystOrdered: CatalystOrdered[A]): CatalystRowOrdered[A] = of[A]
```

Comment:
This might be dangerous if something gets added to `CatalysOrdered` that's not supported in sorting. Does anyone know if there even are any types like that? Would it be safer to explicitly state all the allowable types here even if there is some duplication?

Reply:
If it works nicely with `CatalysOrdered` I would keep it this way for now. Just add a note in `object CatalysOrdered` saying than if someone adds a new case here he should also update the window function tests.

Reply:
Do we need this extra type class? Why wouldn’t Catalyst order cover this case as well?

Reply:
I'm not sure I follow. Are you asking why is `CatalystRowOrdered` needed, shouldn't `CatalystOrdered` be enough? The reason for that is you can order rows by things that aren't comparable within a row, for example Arrays.
```scala
scala> spark.createDataset(Seq((Seq(1, 2, 3), Seq(2,3,1))))
res3: org.apache.spark.sql.Dataset[(Seq[Int], Seq[Int])] = [_1: array<int>, _2: array<int>]

scala> res3.select(untyped.col("_1") > untyped.col("_2")).show
org.apache.spark.sql.AnalysisException: cannot resolve '(`_1` > `_2`)' due to data type mismatch: '(`_1` > `_2`)' requires (boolean or tinyint or smallint or int or bigint or float or double or decimal or timestamp or date or string or binary) type, not array<int>;;

scala> res3.sort(untyped.col("_1")).show()
+---------+---------+
|       _1|       _2|
+---------+---------+
|[1, 2, 3]|[2, 3, 1]|
+---------+---------+
```

Reply:
That looks more like a bug than a feature ...  not yours ofc, on Spark side. Is there any particular reason why they made this design choice?

Reply:
Seems that this is fixed in Spark 2.3: https://issues.apache.org/jira/browse/SPARK-21110. Can you also take a look and verify if this is the same bug. It might save us a lot of extra code if this magically goes away. We can then just focus on one type-class for CatalystOrdered it seems?

Reply:
Good find @imarios! I honestly have no idea if there's a reason for the design choice. I'll grab a more serious look at the fix sometime in the next few days. 
If this does indeed fix everything, are we targeting `2.3` for the next frameless version? Does it make sense to have keep this in for earlier version support?

Reply:
Following up on this. I was able to test the above code snippet ☝️ https://github.com/typelevel/frameless/pull/225/files#r158376192 using spark `2.3.0-SNAPSHOT` and it does indeed work!